tests/test_metadata.py

```diff
@@ -0,0 +1,25 @@
+from wheel.metadata import generate_requirements
+
+
+def test_generate_requirements():
+    extras_require = {
+        'test': ['ipykernel', 'ipython', 'mock'],
+        'test:python_version == "3.3"': ['pytest<3.3.0'],
+        'test:python_version >= "3.4" or python_version == "2.7"': ['pytest'],
+    }
+    expected_metadata = [
+        ('Provides-Extra',
+         'test'),
+        ('Requires-Dist',
+         "ipykernel; extra == 'test'"),
+        ('Requires-Dist',
+         "ipython; extra == 'test'"),
+        ('Requires-Dist',
+         "mock; extra == 'test'"),
+        ('Requires-Dist',
+         'pytest (<3.3.0); (python_version == "3.3") and extra == \'test\''),
+        ('Requires-Dist',
+         'pytest; (python_version >= "3.4" or python_version == "2.7") and extra == \'test\''),
+    ]
+    generated_metadata = sorted(set(generate_requirements(extras_require)))
```

Comment:
One question: why `set()`?

Reply:
The order of the generated metadata is not defined, and the `('Provides-Extra', 'test')` line is present multiple times.

Reply:
That alone seems like a bug. Do you know what causes it?

Reply:
I don't think this is an issue, as `pkg_resources` will correctly handle this case. The code could be amended to:
```diff
 wheel/metadata.py | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git i/wheel/metadata.py w/wheel/metadata.py
index 75dce67..51a8e6f 100644
--- i/wheel/metadata.py
+++ w/wheel/metadata.py
@@ -276,8 +276,9 @@ def pkginfo_to_metadata(egg_info_path, pkginfo_path):
             requires = requires_file.read()
         for extra, reqs in sorted(pkg_resources.split_sections(requires),
                                   key=lambda x: x[0] or ''):
-            for item in generate_requirements({extra: reqs}):
-                pkg_info[item[0]] = item[1]
+            for header, value in generate_requirements({extra: reqs}):
+                if value not in pkg_info.get_all(header, ()):
+                    pkg_info[header] = value
 
     description = pkg_info['Description']
     if description:
```

Reply:
But `generate_requirements()` does not call `pkginfo_to_metadata()` .

Reply:
`generate_requirements` is called multiple times by `pkginfo_to_metadata`, so removing duplicate must be done there, or the code of the later further amended do only call `generate_requirements` one time. Happy to create a separate PR with a dedicated test for this.

Reply:
Nah, it looks like I will want to refactor this code anyway.